### ISSUE TITLE
Real-tone music WADs (edit: currently FLAC) don't play even with DLL's installed

### ISSUE THREAD

**Issue by Zodomaniac** [fabiangreffrath/crispy-doom#95]
I extracted those music pack DLL's into Crispy Doom folder and tried this OGG music WAD DoomMetalVol4_44100 http://www.mediafire.com/download/txmqmc2ob9cq896
As a result, no music plays.

The stderr.txt file says
Error loading midi: Failed loading libvorbisfile-3.dll: Не найден указанный модуль.
('Module not found' in Russian, I have Russian version of Windows)

Tried to copy the file from Crispy 3.2 directory, it doesn't work either.

**Comment by fabiangreffrath** [comment]
Oh dammit! Turns out I tested music playback only in a MSYS2 shell, which has all the required libraries available in its PATH variable. I will have to release a new music pack soon anyway, because of the typo in the BAT file that you found, and I'll pay attention to have all the necessary libraries included next time.

**Comment by Zodomaniac** [comment]
BTW, will MP3 support be included? I once saw a dll named something like libmpeg.dll back in the day, but now it's not present even in the music pack

**Comment by fabiangreffrath** [comment]
Am Samstag, den 12.03.2016, 01:12 -0800 schrieb Zodomaniac:
> BTW, will MP3 support be included? I once saw a dll named something
> like libmpeg.dll back in the day, but now it's not present even in
> the music pack

It has been replaced by libmad.dll, which means something like MPEG
Audio Decoder, and is the prefered alternative over libsmpeg.dll
nowadays.

 - Fabian

**Comment by fabiangreffrath** [comment]
I have issued another pull request for MSYS2. Once this is merged, I will release a new music pack with fixed DLLs.

**Comment by fabiangreffrath** [comment]
All your issues should get fixed by replacing the files in your Crispy Doom directory with the ones from Crispy Doom Music Pack 2: http://www.greffrath.com/~fabian/crispy-doom-music-pack_2.zip

Please report back if issues still remain.

**Comment by Zodomaniac** [comment]
Reporting in: FLAC files don't play.

**Comment by fabiangreffrath** [comment]
Well, it does here.
Did you replace Crispy's `SDL_mixer.dll` with the one from the Music Pack 2?

**Comment by Zodomaniac** [comment]
Yes, I replaced everything.

>

**Comment by fabiangreffrath** [comment]
Hm, what does `stderr.txt` say?

And, could you please rename the music file to `d_runnin.flac` and load it via `crispy-doom -file d_runnin.flac -warp 1`? Does the music play?

**Comment by Zodomaniac** [comment]
>could you please rename the music file to d_runnin.flac and load it via crispy-doom -file d_runnin.flac -warp 1? Does the music play?

This way the music plays! But when a FLAC file is inside a WAD, it doesn't play, and stderr,txt is not even created!

**Comment by fabiangreffrath** [comment]
But this would mean the DLLs aren't at fault. Could you please upload this WAD file somewhere?

**Comment by Zodomaniac** [comment]
I got it! Some FLAC files play, and some don't. In this WAD, D_DM2TTL plays, but D_COUNTD doesn't. Both files are undamaged and play in SLADE.
http://www.mediafire.com/download/dfg8cwr8dbckrha

**Comment by fabiangreffrath** [comment]
The files are indeed intact.

There is code in `i_sdlmusic.c` that checks if a given lump is a MIDI file. If it isn't, the code assumes that the lump is in MUS format and attempts to convert it to MIDI format. The problem is that the code that does this conversion returns success for some lumps even if they aren't in MUS format at all; one of these is d_countd.lmp from the PWAD that you made available for download.

Once there was a check for the presence of a header that identified MUS lumps,but it wad removed from Choco to make some other lumps play properly:
https://github.com/chocolate-doom/chocolate-doom/commit/784e848b326a6f8b0c1377fc7a1b3c5a945cb009

So, I am currently caught between a rock and a hard place: I could add this check back, knowing that this would break playback of the deca.wad titlescreen music or I could leave it as is, knowing that this breaks some other lumps. Will wait for other ideas... :/

**Comment by plumsinus** [comment]
Idea brainstorm:
* Check music for headers of all acceptable music types: flac, ogg, etc.
* Add the check back and extend it to accept the header from the file at deca.wad. How many other invalid-header MUS files in wads are there? My guess is "not too many."
* More robust checks to see if mus2mid was actually successful in producing a proper MIDI. I haven't looked at the mus2mid code at all and I don't know if this is possible or practical.

**Comment by fabiangreffrath** [comment]
> Add the check back and extend it to accept the header from the file at deca.wad. How many other invalid-header MUS files in wads are there? My guess is "not too many."

The lump is pretty broken as it has replaced the "MUS" header with "PWAD". If I added the check for the proper header back, I wouldn't need any of the other points you made.

On this forum, consensus seems to be to not support this music lump at all because of its brokenness and I can imagine to go the same route with Crispy: http://forum.zdoom.org/viewtopic.php?f=7&t=16307

**Comment by fabiangreffrath** [comment]
There is a check prepared in `mus2mid.c` that is currently disabled by means of a pre-compiler macro. Once I enable it, all music plays fine with the sole exception of D_DM2TTL from deca.wad. I will enable this check in the short term:

```
ifdef CHECK_MUS_HEADER
    // Check MUS header
    if (musfileheader.id[0] != 'M'
     || musfileheader.id[1] != 'U'
     || musfileheader.id[2] != 'S'
     || musfileheader.id[3] != 0x1A)
    {
        return true;
    }
#endif
```
